File: android/app/src/main/java/com/turbometa/rayban/utils/APIKeyManager.kt
package com.turbometa.rayban.utils

import android.content.Context
import android.util.Log
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import com.turbometa.rayban.managers.AlibabaEndpoint
import com.turbometa.rayban.managers.APIProvider
import com.turbometa.rayban.managers.APIProviderManager

/**
 * API Key Manager
 * Secure storage and retrieval of API keys using EncryptedSharedPreferences
 * Supports multiple API providers (Alibaba Dashscope Beijing/Singapore, OpenRouter, Google)
 * 1:1 port from iOS APIKeyManager.swift
 */
class APIKeyManager(context: Context) {

    companion object {
        private const val TAG = "APIKeyManager"
        private const val PREFS_NAME = "turbometa_secure_prefs"

        // Account names for different providers
        private const val KEY_ALIBABA_BEIJING = "alibaba-beijing-api-key"
        private const val KEY_ALIBABA_SINGAPORE = "alibaba-singapore-api-key"
        private const val KEY_OPENROUTER = "openrouter-api-key"
        private const val KEY_GOOGLE = "google-api-key"
        private const val KEY_LEGACY = "qwen_api_key" // For backward compatibility

        // Settings keys
        private const val KEY_AI_MODEL = "ai_model"
        private const val KEY_OUTPUT_LANGUAGE = "output_language"
        private const val KEY_VIDEO_QUALITY = "video_quality"
        private const val KEY_RTMP_URL = "rtmp_url"

        @Volatile
        private var instance: APIKeyManager? = null

        fun getInstance(context: Context): APIKeyManager {
            return instance ?: synchronized(this) {
                instance ?: APIKeyManager(context.applicationContext).also { instance = it }
            }
        }
    }

    private val masterKey = MasterKey.Builder(context)
        .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
        .build()

    private val sharedPreferences = EncryptedSharedPreferences.create(
        context,
        PREFS_NAME,
        masterKey,
        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
    )

    init {
        migrateLegacyKey()
    }

    // MARK: - Migration

    private fun migrateLegacyKey() {
        try {
            // Migrate old qwen key to new Alibaba Beijing format
            val legacyKey = sharedPreferences.getString(KEY_LEGACY, null)
            if (!legacyKey.isNullOrBlank() && sharedPreferences.getString(KEY_ALIBABA_BEIJING, null).isNullOrBlank()) {
                sharedPreferences.edit()
                    .putString(KEY_ALIBABA_BEIJING, legacyKey)
                    .remove(KEY_LEGACY)
                    .apply()
                Log.i(TAG, "Migrated legacy qwen API key to Alibaba Beijing")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Migration error: ${e.message}")
        }
    }

    // MARK: - Provider-specific API Key Management

    fun saveAPIKey(key: String, provider: APIProvider, endpoint: AlibabaEndpoint? = null): Boolean {
        return try {
            if (key.isBlank()) return false
            val accountKey = accountName(provider, endpoint)
            sharedPreferences.edit().putString(accountKey, key).apply()
            true
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save API key: ${e.message}")
            false
        }
    }

    fun getAPIKey(provider: APIProvider, endpoint: AlibabaEndpoint? = null): String? {
        return try {
            val accountKey = accountName(provider, endpoint)
            sharedPreferences.getString(accountKey, null)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get API key: ${e.message}")
            null
        }
    }

    fun deleteAPIKey(provider: APIProvider, endpoint: AlibabaEndpoint? = null): Boolean {
        return try {
            val accountKey = accountName(provider, endpoint)
            sharedPreferences.edit().remove(accountKey).apply()
            true
        } catch (e: Exception) {
            Log.e(TAG, "Failed to delete API key: ${e.message}")
            false
        }
    }

    fun hasAPIKey(provider: APIProvider, endpoint: AlibabaEndpoint? = null): Boolean {
        return !getAPIKey(provider, endpoint).isNullOrBlank()
    }

    // MARK: - Google API Key (for Live AI)

    fun saveGoogleAPIKey(key: String): Boolean {
        return try {
            if (key.isBlank()) return false
            sharedPreferences.edit().putString(KEY_GOOGLE, key).apply()
            true
        } catch (e: Exception) {
            Log.e(TAG, "Failed to save Google API key: ${e.message}")
            false
        }
    }

    fun getGoogleAPIKey(): String? {
        return try {
            sharedPreferences.getString(KEY_GOOGLE, null)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to get Google API key: ${e.message}")
            null
        }
    }

    fun deleteGoogleAPIKey(): Boolean {
        return try {
            sharedPreferences.edit().remove(KEY_GOOGLE).apply()
            true
        } catch (e: Exception) {
            Log.e(TAG, "Failed to delete Google API key: ${e.message}")
            false
        }
    }

    fun hasGoogleAPIKey(): Boolean {
        return !getGoogleAPIKey().isNullOrBlank()
    }

    // MARK: - Backward Compatible Methods (defaults to current provider)

    fun saveAPIKey(key: String): Boolean {
        return saveAPIKey(key, APIProviderManager.staticCurrentProvider, APIProviderManager.staticAlibabaEndpoint)
    }

    fun getAPIKey(): String? {
        return getAPIKey(APIProviderManager.staticCurrentProvider, APIProviderManager.staticAlibabaEndpoint)
    }

    fun deleteAPIKey(): Boolean {
        return deleteAPIKey(APIProviderManager.staticCurrentProvider, APIProviderManager.staticAlibabaEndpoint)
    }

    fun hasAPIKey(): Boolean {
        return hasAPIKey(APIProviderManager.staticCurrentProvider, APIProviderManager.staticAlibabaEndpoint)
    }

    // MARK: - Private Helpers

    private fun accountName(provider: APIProvider, endpoint: AlibabaEndpoint?): String {
        return when (provider) {
            APIProvider.ALIBABA -> {
                val effectiveEndpoint = endpoint ?: APIProviderManager.staticAlibabaEndpoint
                when (effectiveEndpoint) {
                    AlibabaEndpoint.BEIJING -> KEY_ALIBABA_BEIJING
                    AlibabaEndpoint.SINGAPORE -> KEY_ALIBABA_SINGAPORE
                }
            }
            APIProvider.OPENROUTER -> KEY_OPENROUTER
        }
    }

    // MARK: - Settings (non-sensitive data)

    // AI Model
    fun saveAIModel(model: String) {
        sharedPreferences.edit().putString(KEY_AI_MODEL, model).apply()
    }

    fun getAIModel(): String {
        return sharedPreferences.getString(KEY_AI_MODEL, "qwen3-omni-flash-realtime") ?: "qwen3-omni-flash-realtime"
    }

    // Output Language
    fun saveOutputLanguage(language: String) {
        sharedPreferences.edit().putString(KEY_OUTPUT_LANGUAGE, language).apply()
    }

    fun getOutputLanguage(): String {
        return sharedPreferences.getString(KEY_OUTPUT_LANGUAGE, "zh-CN") ?: "zh-CN"
    }

    // Video Quality
    fun saveVideoQuality(quality: String) {
        sharedPreferences.edit().putString(KEY_VIDEO_QUALITY, quality).apply()
    }

    fun getVideoQuality(): String {
        return sharedPreferences.getString(KEY_VIDEO_QUALITY, "MEDIUM") ?: "MEDIUM"
    }

    // RTMP URL
    fun saveRtmpUrl(url: String) {
        sharedPreferences.edit().putString(KEY_RTMP_URL, url).apply()
    }

    fun getRtmpUrl(): String? {
        return sharedPreferences.getString(KEY_RTMP_URL, null)
    }
}

// Available AI models for Live AI
enum class AIModel(val id: String, val displayName: String) {
    // Alibaba Qwen Omni
    QWEN_FLASH_REALTIME("qwen3-omni-flash-realtime", "Qwen3 Omni Flash (Realtime)"),
    QWEN_STANDARD_REALTIME("qwen3-omni-standard-realtime", "Qwen3 Omni Standard (Realtime)"),
    // Google Gemini
    GEMINI_FLASH("gemini-2.0-flash-exp", "Gemini 2.0 Flash")
}

// Available output languages
enum class OutputLanguage(val code: String, val displayName: String, val nativeName: String) {
    CHINESE("zh-CN", "Chinese", "\u4e2d\u6587"),
    ENGLISH("en-US", "English", "English"),
    JAPANESE("ja-JP", "Japanese", "\u65e5\u672c\u8a9e"),
    KOREAN("ko-KR", "Korean", "\ud55c\uad6d\uc5b4"),
    SPANISH("es-ES", "Spanish", "Espa\u00f1ol"),
    FRENCH("fr-FR", "French", "Fran\u00e7ais")
}

// Video quality options
enum class StreamQuality(val id: String, val displayNameResId: Int, val descriptionResId: Int) {
    LOW("LOW", com.turbometa.rayban.R.string.quality_low, com.turbometa.rayban.R.string.quality_low_desc),
    MEDIUM("MEDIUM", com.turbometa.rayban.R.string.quality_medium, com.turbometa.rayban.R.string.quality_medium_desc),
    HIGH("HIGH", com.turbometa.rayban.R.string.quality_high, com.turbometa.rayban.R.string.quality_high_desc)
}

File: CameraAccess/Utils/APIKeyManager.swift
/*
 * API Key Manager
 * Secure storage and retrieval of API keys using Keychain
 * Supports multiple API providers (Alibaba Dashscope, OpenRouter, Google)
 */

import Foundation
import Security

class APIKeyManager {
    static let shared = APIKeyManager()

    private let service = "com.turbometa.apikey"

    // Account names for different providers
    private let alibabaBeijingAccount = "alibaba-beijing-api-key"
    private let alibabaSingaporeAccount = "alibaba-singapore-api-key"
    private let openrouterAccount = "openrouter-api-key"
    private let googleAccount = "google-api-key"
    private let legacyAccount = "qwen-api-key" // For backward compatibility (migrates to Beijing)
    private let legacyAlibabaAccount = "alibaba-api-key" // Old format (migrates to Beijing)

    private init() {
        // Migrate legacy key to new format if needed
        migrateLegacyKey()
    }

    // MARK: - Migration

    private func migrateLegacyKey() {
        // Migrate very old qwen key format
        if let legacyKey = getKey(for: legacyAccount),
           getKey(for: alibabaBeijingAccount) == nil {
            _ = saveKey(legacyKey, for: alibabaBeijingAccount)
            _ = deleteKey(for: legacyAccount)
            print("✅ Migrated legacy qwen API key to Alibaba Beijing")
        }

        // Migrate old alibaba key format (without endpoint)
        if let oldAlibabaKey = getKey(for: legacyAlibabaAccount),
           getKey(for: alibabaBeijingAccount) == nil {
            _ = saveKey(oldAlibabaKey, for: alibabaBeijingAccount)
            _ = deleteKey(for: legacyAlibabaAccount)
            print("✅ Migrated old Alibaba API key to Beijing endpoint")
        }
    }

    // MARK: - Provider-specific API Key Management

    func saveAPIKey(_ key: String, for provider: APIProvider, endpoint: AlibabaEndpoint? = nil) -> Bool {
        let account = accountName(for: provider, endpoint: endpoint)
        return saveKey(key, for: account)
    }

    func getAPIKey(for provider: APIProvider, endpoint: AlibabaEndpoint? = nil) -> String? {
        let account = accountName(for: provider, endpoint: endpoint)
        return getKey(for: account)
    }

    func deleteAPIKey(for provider: APIProvider, endpoint: AlibabaEndpoint? = nil) -> Bool {
        let account = accountName(for: provider, endpoint: endpoint)
        return deleteKey(for: account)
    }

    func hasAPIKey(for provider: APIProvider, endpoint: AlibabaEndpoint? = nil) -> Bool {
        return getAPIKey(for: provider, endpoint: endpoint) != nil
    }

    // MARK: - Google API Key (for Live AI)

    func saveGoogleAPIKey(_ key: String) -> Bool {
        return saveKey(key, for: googleAccount)
    }

    func getGoogleAPIKey() -> String? {
        return getKey(for: googleAccount)
    }

    func deleteGoogleAPIKey() -> Bool {
        return deleteKey(for: googleAccount)
    }

    func hasGoogleAPIKey() -> Bool {
        return getGoogleAPIKey() != nil
    }

    // MARK: - Backward Compatible Methods (defaults to current provider)

    func saveAPIKey(_ key: String) -> Bool {
        return saveAPIKey(key, for: APIProviderManager.staticCurrentProvider)
    }

    func getAPIKey() -> String? {
        return getAPIKey(for: APIProviderManager.staticCurrentProvider)
    }

    @discardableResult
    func deleteAPIKey() -> Bool {
        return deleteAPIKey(for: APIProviderManager.staticCurrentProvider)
    }

    func hasAPIKey() -> Bool {
        return hasAPIKey(for: APIProviderManager.staticCurrentProvider)
    }

    // MARK: - Private Helpers

    private func accountName(for provider: APIProvider, endpoint: AlibabaEndpoint? = nil) -> String {
        switch provider {
        case .alibaba:
            // Use current endpoint from settings if not specified
            let effectiveEndpoint = endpoint ?? APIProviderManager.staticAlibabaEndpoint
            switch effectiveEndpoint {
            case .beijing:
                return alibabaBeijingAccount
            case .singapore:
                return alibabaSingaporeAccount
            }
        case .openrouter:
            return openrouterAccount
        }
    }

    private func saveKey(_ key: String, for account: String) -> Bool {
        guard !key.isEmpty else { return false }

        let data = key.data(using: .utf8)!

        // Delete existing key first
        _ = deleteKey(for: account)

        // Add new key
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecValueData as String: data
        ]

        let status = SecItemAdd(query as CFDictionary, nil)
        return status == errSecSuccess
    }

    private func getKey(for account: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne
        ]

        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)

        guard status == errSecSuccess,
              let data = result as? Data,
              let key = String(data: data, encoding: .utf8) else {
            return nil
        }

        return key
    }

    private func deleteKey(for account: String) -> Bool {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account
        ]

        let status = SecItemDelete(query as CFDictionary)
        return status == errSecSuccess || status == errSecItemNotFound
    }
}
